File: python/P209.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        '''Two pointer
        if not nums:
            return 0
        sum_array = [0 for i in range(len(nums)+1)]
        for i in range(len(nums)):
            sum_array[i+1] = sum_array[i]+nums[i]
        if sum_array[-1] < s:
            return 0
        start = 0
        end = 0
        count = len(nums)+100
        while start <= end and end <= len(sum_array)-1:
            if sum_array[end]-sum_array[start] >= s:
                count = min(count, end-start)
                start += 1
            else:
                end += 1
        return count
        '''
        # Binary Search
        if not nums:
            return 0
        sum_array = [0 for i in range(len(nums)+1)]
        for i in range(len(nums)):
            sum_array[i+1] = sum_array[i]+nums[i]
        if sum_array[-1] < s:
            return 0
        answer = len(nums)+1
        left = 0
        for i in range(1, len(sum_array)):
            if sum_array[i] >= s:
                right = i
                left = self.findleft(left, right, sum_array, sum_array[i], s)
                answer = min(answer, right-left+1)
        return answer
            
    def findleft(self, left, right, sum_array, num, s):
        while left < right:
            mid = (left+right)/2
            if num-sum_array[mid] >= s:
                left = mid + 1
            else:
                right = mid
        return left
                

File: java/P209.java
class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        return solvefunction(s, nums);
    }
    
    public int solvefunction(int s, int[] nums) {
        int[] sums = new int[nums.length+1];
        for (int i = 1; i < sums.length; i++) {
            sums[i] = sums[i-1]+nums[i-1];
        }
        int minimum = Integer.MAX_VALUE;
        for (int i = 0; i < sums.length; i++) {
            int end = binarysearch(i+1, sums.length-1, sums[i]+s, sums);
            if (end == sums.length) break;
            if (end-i < minimum) minimum = end-i;
        }
        return minimum == Integer.MAX_VALUE ? 0:minimum;
    }
    
    public int binarysearch(int left, int right, int target, int[] sums) {
        while (left <= right) {
            int mid = left+(right-left)/2;
            if (sums[mid] >= target) right = mid-1;
            else left = mid+1;
        }
        return left;
    }
}